src/vs/workbench/browser/parts/editor/editorPart.ts

```diff
@@ -149,6 +152,55 @@ export class EditorPart extends Part implements IEditorGroupsService, IEditorGro
 		super(Parts.EDITOR_PART, { hasTitle: false }, themeService, storageService, layoutService);
 
 		this.registerListeners();
+
+		this._register(EditSessionRegistry.registerEditSessionsContribution('workbenchEditorLayout', this));
+	}
+
+	private static readonly EditSessionContributionSchemaVersion = 1;
+
+	getStateToStore() {
+		return {
+			version: EditorPart.EditSessionContributionSchemaVersion,
+			serializedGrid: this.gridWidget.serialize(),
+			activeGroup: this._activeGroup.id,
+			mostRecentActiveGroups: this.mostRecentActiveGroups
+		};
+	}
+
+	resumeState(state: unknown, uriResolver: (uri: URI) => URI) {
```

Comment:
This looks to me like a big hack to resume state, in fact you even execute `workbench.action.closeAllEditors` which will result in flicker and would not even close dirty editors. We have a dedicated location on startup where we determine which editors to open and I think we need to integrate session resume there: https://github.com/microsoft/vscode/blob/cca4052e1333f968f313d10f6a10ce322731d420/src/vs/workbench/browser/layout.ts#L756

Having the editors to open there ensures that there will be no flicker and it will also not require to drop the current editor state and recreate it.

Reply:
The issue here is that at startup, the edit session payload is not guaranteed to be available (unless we move towards blocking resuming editor state on a network call to the storage server to retrieve the payload). Moreover an edit session payload today may be applied even after editor startup via the `Resume Latest Changes From Cloud` command, which together with open editors would support scenarios like `https://github.com/microsoft/vscode/issues/35307`.

Reply:
But we cannot just deserialize some state over existing state because you may have dirty editors opened that you cannot just close. So if we want to restore editors, it has to go through a different model that does not drop the grid and creates a new grid.